[AIRFLOW-1349] Fix backfill to respect limits

edgarRd opened the pull request:
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!


### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1349


### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

Before, if a backfill job was triggered that would include a dag run
already in a RUNNING state, the dag run within the backfill would be
included in the count agains the max_active_runs limit. Also, if a
backfill job generated multiple dag runs it could potentially
violate max_active_runs limits by executing all dag runs.

Now the limit is checked per dag run to be created, and the backfill job
will only run the dag runs within the backfill job that could be
included within the limits.
Also, if the max_active_runs limit has already been reached, the
BackfillJob will wait and loop trying to create the required dag runs as
soon as a dag run slot within the limit is available until all dag runs
are completed.

These changes provide a more consistent behavior according to the
max_active_runs limits definition and allows the user to run backfill
jobs with existing RUNNING state when already considered within the
limits.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
  - `BackfillJobTest.test_backfill_max_limit_check*`

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

saguziel commented on airflow/jobs.py:
```diff
@@ -1790,6 +1790,43 @@ class BackfillJob(BaseJob):
+                     requested_runs=0
+                     ):
+            """
+            :param started: Number of tasks already started
```
also put a type annotation

aoen replied:
Good idea factoring this out. 

Started is a dict, let's make the param comment reflect this, same for the other params (they are sets, not numbers).

saguziel commented on airflow/jobs.py:
```diff
@@ -1790,6 +1790,43 @@ class BackfillJob(BaseJob):
 
+    class _Status(object):
+        """
+        Internal status of the backfill job.
```
probably more description

saguziel commented on airflow/jobs.py:
```diff
@@ -1814,40 +1851,37 @@ def __init__(
-        :param succeeded:
-        :param skipped:
-        :param failed:
+        :param status:
```
also type

aoen commented on airflow/jobs.py:
```diff
@@ -1790,6 +1790,43 @@ class BackfillJob(BaseJob):
         'polymorphic_identity': 'BackfillJob'
     }
 
+    class _Status(object):
```
Let's pick a more verbose name so it's clear what this is (Status sounds like the status of the backfill itself). Something like _TIStatuses . Same thing for the params called "status" used in other functions

aoen commented on airflow/jobs.py:
```diff
@@ -1891,118 +1925,92 @@ def _manage_executor_state(self, started):
 
-    def _execute(self):
+    @provide_session
+    def _tasks_instances_for_run_date(self, run_date, session=None):
```
I think returning two things is a little bit unclean here. I think it would be fine to just duplicate the query to get the dagrun in the caller function if it's necessary. Alternatively we could make dagrun an optional parameter for this function (so we would only compute dagrun if the caller does not specify it), but I think performance isn't that important for backfill jobs since they are adhoc (except for subdag operators which are going to go away eventually).

Also "tasks_instances" should be "task_instances"

edgarRd replied:
I thought it'd be better to actually split this function into two functions, preserving the way queries are executed, avoiding tuple return values.

aoen commented on airflow/models.py:
```diff
@@ -3024,6 +3024,24 @@ def get_active_runs(self, session=None):
         return active_dates
 
     @provide_session
+    def count_active_runs(self, external_trigger=None, session=None):
```
get_num_active_runs is more standard also would be good to replace the other places that used the old query with this new function too:
/Users/dan_davydov/ws/airflow/airflow/ti_deps/deps/dagrun_exists_dep.py

edgarRd replied:
Got a comment from @bolkedebruin that this would increase the scope of this PR, so punting to other refactor.

aoen commented on airflow/jobs.py:
```diff
@@ -2193,19 +2191,121 @@ def _execute(self):
+
+        return err
+
+    def _get_run_dates(self):
```
It would be good to add tests for this function.

edgarRd replied:
Added tests.

aoen reviewed (commented):
It's a little bit tough to review this PR, can we split this PR up into the refactoring bits first and then the semantic change/bug fix afterwards?

Also, looks like there are some new codequality issues: https://landscape.io/github/apache/incubator-airflow/pulls/2454

bolkedebruin commented on airflow/models.py:
```diff
@@ -3024,6 +3024,25 @@ def get_active_runs(self, session=None):
         return active_dates
 
     @provide_session
+    def get_num_active_runs(self, external_trigger=None, session=None):
```
why not extend get_active_runs a little bit and do a len(get_active_runs) or just use DagRun.find? I don't see the reason for this function.

edgarRd replied:
I thought about that option, but the following reasons made me not taking that route:
1. `get_active_runs` is actually inconsistent - the name does not reflect what it does, hence it needs to be changed; additionally, it's not very useful as it only return execution_dates instead of dag run objects to be more generic and actually use for this purpose as well.
2. Extending `get_active_runs` a little bit would require more thought since the implementation seems very particular to a use case, again, returning only the execution_dates. Thus, this will broad the scope of the PR, which we're trying to avoid.
3. Even though performance seems not to be an issue in this case, having a query that only fetched the numeric count would reduce the number of bits going through the pipe, so that's always good. There are many cases where we don't need the full payload of all active runs, specially when calling the count repeatedly.

Hence, I decided to go against using `get_active_runs` and implement a more sensitive function: `get_num_active_runs` (renamed after suggestion by @aoen) - Also, seems like this same query could be used in some other parts of the code base, but I think for now to avoid further changes in other places we can avoid them.

bolkedebruin replied:
Are you sure your count() function is faster: http://docs.sqlalchemy.org/en/latest/orm/query.html and https://stackoverflow.com/questions/14754994/why-is-sqlalchemy-count-much-slower-than-the-raw-query ?

As performance indeed isn't a real issue. How many active dag runs will you find? 200? 400? 1000? I'm not particularly convinced it warrants the extra complexity of adding this function or not using DagRun.find.

edgarRd replied:
> Are you sure your count() function is faster: http://docs.sqlalchemy.org/en/latest/orm/query.html and https://stackoverflow.com/questions/14754994/why-is-sqlalchemy-count-much-slower-than-the-raw-query ?

I'm not claiming that the `count()` query would be faster, furthermore the performance of the query does not even depend on sqlalchemy but the query engine with the optimizations that it can do. This does reduce the payload on transfer.

> As performance indeed isn't a real issue. How many active dag runs will you find? 200? 400? 1000? I'm not particularly convinced it warrants the extra complexity of adding this function or not using DagRun.find.

Like you mentioned, query performance is not an issue, but this does reduce the payload transferred. I think this function does not add any more complexity than the one is already there. Additionally, like I mentioned before, it seems this method could be reused in other bits of the codebase; using `DagRun.find` would just duplicate code on every place this can be used.

saguziel commented on airflow/jobs.py:
```diff
@@ -1790,6 +1790,58 @@ class BackfillJob(BaseJob):
         'polymorphic_identity': 'BackfillJob'
     }
 
+    class _TiStatus(object):
```
sounds like it is task status? TI is defined by task (task_id + dag_id) and execution_date

edgarRd committed:
[AIRFLOW-1349] Fix backfill to respect limits

Before, if a backfill job was triggered that would include a dag run
already in a RUNNING state, the dag run within the backfill would be
included in the count agains the max_active_runs limit. Also, if a
backfill job generated multiple dag runs it could potentially
violate max_active_runs limits by executing all dag runs.

Now the limit is checked per dag run to be created, and the backfill job
will only run the dag runs within the backfill job that could be
included within the limits.
Also, if the max_active_runs limit has already been reached, the
BackfillJob will wait and loop trying to create the required dag runs as
soon as a dag run slot within the limit is available until all dag runs
are completed.

These changes provide a more consistent behavior according to the
max_active_runs limits definition and allows the user to run backfill
jobs with existing RUNNING state when already considered within the
limits.

aoen commented on airflow/jobs.py:
```diff
@@ -1857,6 +1858,8 @@ def __init__(self,
             :param active_runs: Active tasks at a certain point in time
             :type active_runs: list
+            :param executed_dag_run_dates: Datetime objects for the executed dag runs
+            :type executed_dag_run_dates: set
```
Needs a type set(TYPE)

aoen commented on airflow/jobs.py:
```diff
@@ -1975,22 +1979,36 @@ def _manage_executor_state(self, started):
         """
         Returns a dag run for the given run date, which will be matched to an existing
-        dag run if available or create a new dag run otherwise.
+        dag run if available or create a new dag run otherwise. The returned dag run
```
returned dag run should be created dag run right (i.e. it's possible to return a dagrun with state == failed if it already existed to begin with). Same comment for the return annotation below.

aoen commented on airflow/jobs.py:
```diff
@@ -1975,22 +1979,36 @@ def _manage_executor_state(self, started):
 
+        # consider max_active_runs but ignore when running subdags
+        # "parent.child" as a dag_id is by convention a subdag
+        max_active_limit = (self.dag.max_active_runs
```
I think boolean is a bit cleaner/explicit here than setting the int to max value as a way to store the additional information.
i.e. respect_max_active_limit = True/False and if it's True set max_active_limit. Then below you can check "if respect and current_active_dag_count >= max_active_limit"

aoen commented on airflow/jobs.py:
```diff
@@ -1975,22 +1979,36 @@ def _manage_executor_state(self, started):
 
         if run is None or len(run) == 0:
+            if current_active_dag_count >= max_active_limit:
+                return no_dag_run
```
I think just returning None is fine, otherwise should make this a class constant

aoen commented on airflow/jobs.py:
```diff
@@ -2333,22 +2365,7 @@ def _execute(self):
-                self.logger.info("Dag {} has reached maximum amount of {} dag runs"
-                                 .format(self.dag.dag_id, self.dag.max_active_runs))
-                return
+        delay_in_secs = 1
```
Let's make this configurable.

edgarRd replied:
As a CLI parameter?

aoen replied:
yeah that works.

aoen commented on airflow/jobs.py:
```diff
@@ -2372,20 +2389,33 @@ def _execute(self):
+                                            start_date=start_date,
+                                            session=session)
+
+                run_dates = [run_date for run_date in run_dates
```
Let's use a different variable rather than mutating the same one. It looks like you're only checking the length anyways though so maybe just replace the len(run_dates) > 0 below with an:
any(run_date not in ti_status.executed_dag_run_dates for run_date in run_dates) and remove this variable assignment

edgarRd replied:
The `while` check with `any(run_date not in ti_status.executed_dag_run_dates for run_date in run_dates)` would not be enough, we'd still need to iterate the `run_dates` again to pass the remaining dates to `self._execute_for_run_dates`, otherwise we'll keep passing all the run_dates even the ones already processed - originally I mutated the same variable as a way to filter out the dates already processed and avoiding doing the same iteration of dates multiple times. 
I could replace the mutation of `run_dates` by a control `int` variable, effectively a counter, that keeps the number of remaining tasks and gets updated after processing the dates by doing: `len(run_dates) - len(ti_status.executed_dag_run_dates)`.

aoen replied:
Sorry I meant the second run_dates assignment (the one just above this comment) could just be replaced with the any check in the if statement below. The original one can stay intact.

aoen replied:
Spoke offline, didn't notice the while loop!

aoen commented on tests/jobs.py:
```diff
@@ -339,22 +341,79 @@ def test_backfill_max_limit_check_within_limit(self):
+        # at this point backfill can't run since the max_active_runs has been reached,
+        # so it is waiting
+        dagruns = None
+        while not dagruns:
```
Get rid of the sleep if possible, it's an antipattern in tests, haven't looked at this this specific case but usually you want to receive a push notification from the source. At the least we need a timeout here.

edgarRd replied:
Yeah, I don't think we really need the sleep here.

aoen replied:
Should have been a bit more clear sorry, the while loop is a problem too, typically this is solved by expecting some response and timing out if it takes more than X seconds (difference is it's a push, not a pull/poll). If it's too hard to set this up we can put in a sleep for now (ideally a smaller one, and have a timeout on the while loop to make sure the test suite doesn't crash/stall on this single test.

edgarRd replied:
Right, basically a callback. I think the easiest would be using a count down latch or a thread condition with a timeout.

aoen commented on airflow/jobs.py:
```diff
@@ -1975,22 +1979,36 @@ def _manage_executor_state(self, started):
         run_id = BackfillJob.ID_FORMAT_PREFIX.format(run_date.isoformat())
 
+        # consider max_active_runs but ignore when running subdags
+        # "parent.child" as a dag_id is by convention a subdag
```
Nit: is this second comment needed since we are calling the is_subdag method anyways and the implementation isn't relevant to this function?

aoen commented on airflow/jobs.py:
```diff
@@ -2001,6 +2019,12 @@ def _get_dag_run(self, run_date, session=None):
+            if run.state == State.SUCCESS:
+                return run
+
+            if (run.state != State.RUNNING and
```
Discussed offline, we are going to remove success as a special case dagrun state here to make this function simpler.

aoen commented on airflow/models.py:
```diff
@@ -3524,9 +3543,34 @@ def run(
             ignore_first_depends_on_past=False,
-            pool=None):
+            pool=None,
+            delay_on_limit_secs=1.0):
```
Let's keep the default value in only one place (1.0), can this be a required arg instead (remove =1.0)?

aoen replied:
Also same comment for the other places this is done. LGTM after this.

edgarRd replied:
I'm not sure about making it a positional argument, it would need to be included in all other places where `BackfillJob` is already instantiated in the codebase, and also in the future. Furthermore, I'm not sure if it's strictly required, since the underlying code won't always use it, as long as there's a sensible default for when it is needed. 

Ideally you'd have a `defaults` module (I guess `settings` would be similar) in a top package where all these values can live in one place and just make a reference to everywhere else in the codebase if needed, given that we have the same issue with all other optional args in this function and others similar. I could change this to a similar approach.

aoen replied:
Alright not a huge deal, I'll merge now.
